pkg/ingester/limiter.go

```diff
@@ -210,34 +211,55 @@ func (l *Limiter) convertGlobalToLocalLimit(userID string, globalLimit int) int
 		return 0
 	}
 
-	// Given we don't need a super accurate count (ie. when the ingesters
-	// topology changes) and we prefer to always be in favor of the tenant,
-	// we can use a per-ingester limit equal to:
-	// (global limit / number of ingesters) * replication factor
-	numIngesters := l.ring.HealthyInstancesCount()
+	healthyIngestersCount := l.ring.HealthyInstancesCount()
 
 	// May happen because the number of ingesters is asynchronously updated.
 	// If happens, we just temporarily ignore the global limit.
-	if numIngesters == 0 {
+	if healthyIngestersCount == 0 {
 		return 0
 	}
 
-	// If the number of available ingesters is greater than the tenant's shard
-	// size, then we should honor the shard size because series/metadata won't
-	// be written to more ingesters than it.
-	if shardSize := l.getShardSize(userID); shardSize > 0 {
-		// We use Min() to protect from the case the expected shard size is > available ingesters.
-		numIngesters = util_math.Min(numIngesters, util.ShuffleShardExpectedInstances(shardSize, l.getNumZones()))
+	zonesCount := l.getZonesCount()
+	shardSize := l.getShardSize(userID)
+	var ingestersInZoneCount int
+	if zonesCount > 1 {
+		// In this case zone-aware replication is enabled, and ingestersInZoneCount is initially set to
+		// the exact number of healthy ingesters in the corresponding zone
+		ingestersInZoneCount = l.ring.HealthyInstancesInZoneCount()
+		if shardSize > 0 {
+			// If shuffle sharding is enabled and the number of healthy ingesters in the zone is greater than the
+			// expected number of ingesters per sharded zone, then we should honor the latter because series/metadata
+			// cannot be written to more ingesters than that.
+			ingestersInZoneCount = util_math.Min(ingestersInZoneCount, util.ShuffleShardExpectedInstancesPerZone(shardSize, zonesCount))
+		}
+	} else {
+		// In this case zone-aware replication is disabled, and ingestersInZoneCount is initially set to
+		// the exact numnber of healthy ingesters
+		ingestersInZoneCount = healthyIngestersCount
+		if shardSize > 0 {
+			// If shuffle sharding is enabled, and the number of healthy ingesters is greater than the tenant's shard size,
+			// then we should honor the latter because series/metadata cannot be written to more ingesters than that.
+			ingestersInZoneCount = util_math.Min(healthyIngestersCount, util.ShuffleShardExpectedInstances(shardSize, zonesCount))
+		}
 	}
 
-	return int((float64(globalLimit) / float64(numIngesters)) * float64(l.replicationFactor))
+	// This may happen, for example when zone-aware replication is enabled,
+	// and all ingesters in a zone have been scaled down
+	if ingestersInZoneCount == 0 {
+		return 0
+	}
+
+	// Global limit is equally distributed among all the active zones.
+	// The portion of global limit related to each zone is then equally distributed
+	// among all the ingesters belonging to that zoneInside of each zone,
```

Comment:
[nit] I think there's a typo here: "zoneInside" and "," at the end of the sentence.